### ISSUE TITLE
Enable code coverage Github Actions

### ISSUE THREAD

**Issue by Skylion007** [sympy/sympy#24831]
Since we are now on Github Actions, it would be nice if we enabled automated code coverage reports for PRs to find bugs where test cases aren't actually testing the right functionality etc. Codecov.io is a popular website to upload code coverage reports and has good Github action integration (https://github.com/codecov/codecov-action), but there are various other tools that can be used instead. We already have scripts that generate the codecov reports: https://github.com/sympy/sympy/blob/c7fe006d84d846a548e17aa4707e3639573b7a51/bin/coverage_report.py so we might as well display it on PR and show the diff in coverage that each PR is adding to the codebase.

**Comment by oscarbenjamin** [comment]
SymPy did use codecov before but the coverage testing was not migrated to GitHub Actions. I think it is generally agreed that something like it would be good (see discussion at gh-21328).

The first thing that is needed is to add coverage jobs to CI that actually measure coverage while running the tests so that codecov can upload those files. If you are interested to test this out I suggest enabling codecov in your fork so that you can then demonstrate it. Then you could open a PR with the changes needed for it to work in the SymPy repo.

**Comment by sylee957** [comment]
I'm not sure about why the codecov didn't get migrated to github actions.
But you can browse the history of SymPy and reference the code coverage stuff we used to work for Travis CI to get it back.

**Comment by oscarbenjamin** [comment]
> I'm not sure about why the codecov didn't get migrated to github actions.

It was because migrating to GitHub Actions was a ton of work that no one else helped me with at the time. I migrated the essentials to keep the repo working  but that was already more work than I wanted it to be so I left the rest for others to pick up on. Since no one else decided to migrate the coverage testing it didn't get migrated.

**Comment by janujan99** [comment]
Hi, I'd be happy to work on this if no one else is doing so currently. I was looking for issues to contribute to as part of my GSOC application, and I have some experience with setting up code coverage.

**Comment by victorian177** [comment]
Hey, @oscarbenjamin. I have written a code coverage workflow [here]
(https://github.com/victorian177/sympy/blob/coverage_workflow/.github/workflows/coverage.yaml). This is my first time writing a GItHub workflow report. I would like to know what I may need to change.

**Comment by oscarbenjamin** [comment]
This should probably be a job that runs in the main `runtests` workflow rather than a separate workflow. There is no point running the coverage measurement if the tests don't pass.

> I would like to know what I may need to change.

Does it work? Do you see anything from codecov?

**Comment by victorian177** [comment]
Hey, @oscarbenjamin
> Does it work? Do you see anything from codecov?
Not yet. I have added the coverage code to the runtests workflow file. I left out the codecov part. Wanted to ensure that the coverage part is working as expected and then I would add the codecov.

**Comment by victorian177** [comment]
@oscarbenjamin. I have added the coverage job to the [runtests workflow](https://github.com/victorian177/sympy/blob/coverage_workflow/.github/workflows/runtests.yml#L503-L516), running the coverage_report.py. I am yet to add the uploading to codecov. Wanted to ensure it works as should before I add uploading to codecov. I encountered errors when running the workflow using [coverage_report.py](https://github.com/victorian177/sympy/blob/master/bin/coverage_report.py) module on [GitHub Actions](https://github.com/victorian177/sympy/actions/runs/4415860742). I would run it locally look to see if the errors persists and fix them.

**Comment by oscarbenjamin** [comment]
A particularly useful thing would be to check diff-cover and report a failing job if the lines in the diff are not covered by tests. (It could be a not required job.)

**Comment by asmeurer** [comment]
Should we run coverage in a completely separate job or just enable it in one of the main jobs? Enabling it does make the tests run slower (I don't know if it's by enough that we have to do more splits). But running a duplicate job seems redundant.

We would also probably want to enable coverage in one of the optional dependencies builds and merge it or else we won't have any coverage data for optional dependency code. That could come later though as merging reports across different jobs is harder.

**Comment by oscarbenjamin** [comment]
> I don't know if it's by enough that we have to do more splits

It certainly used to be. The coverage tests had 4 splits compared to 2 because they were so much slower.

**Comment by asmeurer** [comment]
GitHub Actions runs faster and has a longer time limit than Travis. I guess we'll just have to see how long it takes.

**Comment by victorian177** [comment]
@oscarbenjamin, @asmeurer 

> @oscarbenjamin. I have added the coverage job to the [runtests workflow](https://github.com/victorian177/sympy/blob/coverage_workflow/.github/workflows/runtests.yml#L503-L516), running the coverage_report.py. I am yet to add the uploading to codecov. Wanted to ensure it works as should before I add uploading to codecov. I encountered errors when running the workflow using [coverage_report.py](https://github.com/victorian177/sympy/blob/master/bin/coverage_report.py) module on [GitHub Actions](https://github.com/victorian177/sympy/actions/runs/4415860742). I would run it locally look to see if the errors persists and fix them.

I have been able to successfully run coverage_report.py to generate [coverage reports](https://github.com/victorian177/sympy/actions/runs/4438450192). Using Python version 3.11 led to a RecursionError and that is probably due to 3.11's implementation of a standard library ast.py. Changing the version to 3.9 enabled the coverage job to run successfully.

**Comment by Skylion007** [comment]
Wait, doesn't https://github.com/sympy/sympy/pull/24291 mean we can just install pytest-cov, now and have a working coverage report on CI? https://pypi.org/project/pytest-cov/

**Comment by oscarbenjamin** [comment]
It has always been possible to use `pytest-cov`. Currently none of the jobs runs with coverage measurement enabled though so it would be enabled for one particular Python version.

**Comment by Skylion007** [comment]
codecov can merge together coverage reports from lots of different runs so you could run it on every Python version, if you wanted to. Seems excessive though.

**Comment by oscarbenjamin** [comment]
> Seems excessive though.

It is definitely excessive. It doubles the runtime for the already very slow tests. I would be in favour of reducing the number of Python versions that we test altogether. Certainly coverage measurement does not need to be duplicated.

**Comment by victorian177** [comment]
> Wait, doesn't #24291 mean we can just install [pytest-cov](https://pypi.org/project/pytest-cov/), now and have a working coverage report on CI? Or we can just do `coverage run -m pytest ...` now

Using the [pytest-cov](https://github.com/victorian177/sympy/actions/runs/4447698367/jobs/7810105582) uses way more time than [coverage_report](https://github.com/victorian177/sympy/actions/runs/4438450192/jobs/7790271767).py (over 5 hrs for pytest-cov compared to 1hr30 for coverage_report.py). I am investigating why that is the case. Those are the latest workflow runs. The full list is [here](https://github.com/victorian177/sympy/actions).

**Comment by oscarbenjamin** [comment]
> over 5 hrs for pytest-cov compared to 1hr30 for coverage_report.py

Is that running the slow tests as well as the not slow tests?

Make sure you are using latest master for this after gh-24291 was merged because that changes the behaviour of pytest to skip the slow tests by default.

Also you could try using `--assert=plain` with pytest.

Otherwise I don't think that there should be any significant difference between pytest-cov and bin/coverage_report.py.

**Comment by victorian177** [comment]
Now pytest-cov seems to run in due time now.

**Comment by oscarbenjamin** [comment]
> Now pytest-cov seems to run in due time now.

What exactly has changed? Is it just not running the slow tests?

**Comment by SaxdaAnhalta** [comment]
Could I give this another try with a new PR? #24971 stalled on setup problems back then. I would use the codecov GitHub action unless a different tool is preferred.

**Comment by oscarbenjamin** [comment]
Yes, you can test this out with a PR in your own fork though because you won't have permissions to add Codecov.io in this repo. I'm not sure whether there are better options than codecov now.

**Comment by Skylion007** [comment]
Other option would be using some of Github native codecov tracking or having a bot that leaves a comment on the PR comparing coverage
